### PR TITLE
[SR-522][SR-629][SR-4161] Allow covariance in protocol conformance.

### PR DESCRIPTION
**Pull request by aleksgapp**
I'd like to show you my naive attempt to allow covariance in protocol conformance.
I'm completely unsure if this is an acceptable solution, but it works 😅 

Let me know what do you think of it?

Resolves [SR-522](https://bugs.swift.org/browse/SR-522), [SR-629](https://bugs.swift.org/browse/SR-629), [SR-4161](https://bugs.swift.org/browse/SR-4161).

### CODE REVIEW

File: lib/Sema/TypeCheckProtocol.cpp [hunk 1]
```diff
@@ -1255,9 +1252,16 @@ matchWitness(TypeChecker &tc,
...
 
   // Match a type in the requirement to a type in the witness.
-  auto matchTypes = [&](Type reqType, Type witnessType) 
+  auto matchTypes = [&](Type reqType, Type witnessType)
                       -> Optional<RequirementMatch> {
-    cs->addConstraint(ConstraintKind::Equal, reqType, witnessType, locator);
+    ConstraintKind kind = ConstraintKind::Conversion;
+    if (auto func = dyn_cast<FuncDecl>(req)) {
+      auto name = req->getFullName().getBaseName();
+      if (func->isOperator() && name.str() == "==") {
```
Comment (CodaFi) [inline lib/Sema/TypeCheckProtocol.cpp]: Why is this restriction specifically in place?  Going by name alone isn't going to only pick out Equatable requirements.

File: lib/SILGen/SILGenPoly.cpp [hunk 2]
```diff
@@ -399,6 +399,12 @@ ManagedValue Transform::transform(ManagedValue v,
     inputOTK = OTK_None;
   }
 
+  if (inputOTK != OTK_None && outputOTK == OTK_None) {
+    SILValue result = SGF.B.createUncheckedBitCast(Loc, v.getValue(),
```
Comment (slavapestov) [inline lib/SILGen/SILGenPoly.cpp]: This is not correct -- why are you ending up here?

File: lib/Sema/CSSimplify.cpp [hunk 3]
```diff
@@ -1038,9 +1038,13 @@ ConstraintSystem::matchFunctionTypes(FunctionType *func1, FunctionType *func2,
...
 
-  // Input types can be contravariant (or equal).
-  SolutionKind result = matchTypes(func2->getInput(), func1->getInput(),
-                                   subKind, subflags,
+  // Input types can be contravariant (or equal),
+  // but witness constranit might be covarian (or equal).
+  bool isWitnessConstraint = locator.last() &&
+    locator.last()->getKind() == ConstraintLocator::Witness;
+  Type type1 = isWitnessConstraint ? func1->getInput() : func2->getInput();
+  Type type2 = isWitnessConstraint ? func2->getInput() : func1->getInput();
```
Comment (slavapestov) [inline lib/Sema/CSSimplify.cpp]: The caller should instead call matchTypes() with the right order, instead of flipping them here.

File: lib/Sema/TypeCheckProtocol.cpp [hunk 4]
```diff
@@ -1225,19 +1224,17 @@ matchWitness(TypeChecker &tc,
 
     // Open up the witness type.
     witnessType = witness->getInterfaceType();
-    // FIXME: witness as a base locator?
```
Comment (slavapestov) [inline lib/Sema/TypeCheckProtocol.cpp]: Why did you remove the FIXME?

File: lib/Sema/TypeCheckProtocol.cpp [hunk 5]
```diff
@@ -1255,9 +1252,16 @@ matchWitness(TypeChecker &tc,
...
                       -> Optional<RequirementMatch> {
-    cs->addConstraint(ConstraintKind::Equal, reqType, witnessType, locator);
+    ConstraintKind kind = ConstraintKind::Conversion;
+    if (auto func = dyn_cast<FuncDecl>(req)) {
+      auto name = req->getFullName().getBaseName();
+      if (func->isOperator() && name.str() == "==") {
+        kind = ConstraintKind::Equal;
+      }
+    }
+    cs->addConstraint(kind, witnessType, reqType, locator);
```
Comment (slavapestov) [inline lib/Sema/TypeCheckProtocol.cpp]: Why did you change the order of arguments in the constraint here?

File: lib/Sema/TypeCheckProtocol.cpp [hunk 6]
```diff
@@ -1255,9 +1252,16 @@ matchWitness(TypeChecker &tc,
   };
 
   // Match a type in the requirement to a type in the witness.
-  auto matchTypes = [&](Type reqType, Type witnessType) 
+  auto matchTypes = [&](Type reqType, Type witnessType)
                       -> Optional<RequirementMatch> {
-    cs->addConstraint(ConstraintKind::Equal, reqType, witnessType, locator);
+    ConstraintKind kind = ConstraintKind::Conversion;
```
Comment (slavapestov) [inline lib/Sema/TypeCheckProtocol.cpp]: 'Conversion' is too general, because there are conversions that SILGenPoly cannot express. At most this should be 'Subtype'.

File: test/TypeCoercion/protocol_covariance.swift [hunk 7]
```diff
@@ -0,0 +1,64 @@
+// RUN: %target-typecheck-verify-swift
+
+class A {}
```
Comment (slavapestov) [inline test/TypeCoercion/protocol_covariance.swift]: Please put this test in test/decl/protocol, and add some error cases as well.

Also you need a SILGen test and an executable test.

File: lib/Sema/TypeCheckProtocol.cpp (hunk 1 above)
Comment (aleksgapp) [inline lib/Sema/TypeCheckProtocol.cpp]: There is bunch of errors specifically to `Equtable` conformance after relaxing the constraint due to multiple candidate may exactly match the requirement, I'll add test for this.

I guess you are right, but I coudn't come up with a failing case. Probably we could use `getKnownProtocolKind()` here, but it's doing the same :-/.

File: lib/SILGen/SILGenPoly.cpp (hunk 2 above)
Comment (aleksgapp) [inline lib/SILGen/SILGenPoly.cpp]: `assert` is firing in this function, because requirement is `Optional`, and witness isn't. 

I'd love to avoid ending up here :D

File: lib/Sema/CSSimplify.cpp (hunk 3 above)
Comment (aleksgapp) [inline lib/Sema/CSSimplify.cpp]: 👍

File: lib/Sema/TypeCheckProtocol.cpp (hunk 4 above)
Comment (aleksgapp) [inline lib/Sema/TypeCheckProtocol.cpp]: Previously consranits were added with nullptr locator, I've changed it to witness locator, I thought that's what FIXME was about, am I wrong here?
```cpp
locator = cs->getConstraintLocator(nullptr);
...
cs->addConstraint(ConstraintKind::Equal, reqType, witnessType, locator);
```

File: lib/Sema/TypeCheckProtocol.cpp (hunk 5 above)
Comment (aleksgapp) [inline lib/Sema/TypeCheckProtocol.cpp]: Because relaxing `ConstraintKind` isn't enough.

In this example:
```swift
protocol P {
    var a: Int? { get }
}

struct S: P {
    var a: Int
}
```
Constraint system would try to match `Optional<Int>` (as the requirement) to `Int` (as witness) and it would fail.

In contrast to function type convertion:
```swift
func f() -> Int { return 3 }
let g : ()->Int? = f
```
Constraint system would try to match `Int` to `Optional<Int>`, which works fine.

File: test/TypeCoercion/protocol_covariance.swift (hunk 7 above)
Comment (aleksgapp) [inline test/TypeCoercion/protocol_covariance.swift]: 👍

File: lib/Sema/TypeCheckProtocol.cpp (hunk 6 above)
Comment (aleksgapp) [inline lib/Sema/TypeCheckProtocol.cpp]: Hm, probably that's way I'm getting an assert in `SILGenPoly.cpp`.
I'll see how `Subtype` would work.

File: lib/Sema/TypeCheckProtocol.cpp (hunk 1 above)
Comment (CodaFi) [inline lib/Sema/TypeCheckProtocol.cpp]: Seems like that's your answer then: Variance constraints inside protocols with typealiases or generic requirements in general are going to cause inconsistencies.

Comment (slavapestov) [inline lib/Sema/TypeCheckProtocol.cpp]: Yeah, please add a test so we can take a look. Once we figure out the problem in more detail, we need to solve it in a general way -- special-casing Equatable is not going to work (imagine if the user defines their own protocol with a similar property).

File: lib/SILGen/SILGenPoly.cpp (hunk 2 above)
Comment (slavapestov) [inline lib/SILGen/SILGenPoly.cpp]: If the input is optional and the output is not, you're forcing off a level of optionality. Maybe something is going in the wrong direction?

Also in general, `Optional<T>` is not bitwise-castable to `T` -- this only holds for classes.

File: lib/Sema/TypeCheckProtocol.cpp (hunk 4 above)
Comment (slavapestov) [inline lib/Sema/TypeCheckProtocol.cpp]: Oh I see. I would make the change in an earlier patch so we can test it independently. I think @DougGregor should look at this since I don't understand why we need the locator here.

File: lib/Sema/TypeCheckProtocol.cpp (hunk 5 above)
Comment (slavapestov) [inline lib/Sema/TypeCheckProtocol.cpp]: Oh. So I think the problem is for function arguments, we need to match the witness against the requirement; for results, we need to match the requirement against the witness. This should give you the correct variance and eliminate the need to swap the arguments in matchTypes() above.

**Comment by slavapestov** [comment]
Thanks for working on this! A word of warning -- this is a big change and will require some iteration and a *lot* more tests. :-) So I hope you don't mind some on-going engagement here before we can merge this in.

Other than the immediate review feedback above, a couple of things to consider before we can take this:

- How do we handle ambiguity? With the new approach there might be multiple candidate witnesses for a given requirement. The code in TypeCheckProtocol already has some kind of ranking algorithm but it will need to be generalized.

- What about source compatibility? For example, in the following example, Swift 3 picks the witness from the protocol extension, ignoring the candidate in the struct, because it has the wrong type:

```
protocol P {
  func req() -> String?
}

extension P {
  // default implementation
  func req() -> String? { return nil }
}

// conforming type
struct S : P {
  // almost, but not quite, a match
  func req() -> String { return "hi!") }
}
```

With your change, `S.req()` is now a potential candidate. Are we going to pick it as the witness instead, possibly breaking source compatibility? Or will we pick the extension member which has an exact type?

- In either case, it's probably safest to only execute the new behavior in Swift 4 mode. In fact, until this is fully debugged, I would add a staging flag (see how I did it in https://github.com/apple/swift/commit/de323b5befc664026999b9f18b0af06ebd24cad7 for an example).

- Do we want different rules for `@objc` and native protocols? Right now we have different rules -- `@objc` protocols can change optionality between the requirement and the witness, and native ones cannot. This is inconsistent and the optionality matching is implemented in a bespoke way, and not with a `Subtype` constraint. If you think of a way to unify this it would be great.

- Do we want to extend the new rules to method overrides as well? Again, method override matching uses its own logic and not a constraint solver. It would be awesome to move it over to the solver also (but again you'll have Swift 3 vs 4 compatibility concerns to resolve).

File: lib/SILGen/SILGenPoly.cpp (hunk 2 above)
Comment (aleksgapp) [inline lib/SILGen/SILGenPoly.cpp]: Probably, anyway there's more ambiguity issues after switching constraint to `Subtype`, it will take me some time to figure this out, I'll make an update when ready.

File: lib/Sema/TypeCheckProtocol.cpp (hunk 5 above)
Comment (aleksgapp) [inline lib/Sema/TypeCheckProtocol.cpp]: :-/ Why? I thought to get covariant behaviour we should always match witness agains requirement.

Also, `matchFunctionTypes` flips function arguments itself, as comment states:
> // Input types can be contravariant (or equal)

but can they? Seems there is some inconsistency:

```swift
func subtypeArgument1(_ x: (_ fn: ((String) -> Int)) -> Int) { }
func testSubtypeArgument1(x2: (_ fn: ((String) throws -> Int)) -> Int) {
  subtypeArgument1(x2)
}
```
This works fine.

```swift
func subtypeArgument1(_ x:(Int) -> Int) {}
func testSubtypeArgument1(x2: (Int) throws -> Int) {
  subtypeArgument1(x2)
}
```
This gives an error:
> error: invalid conversion from throwing function of type '(Int) throws -> Int' to non-throwing function type '(Int) -> Int'

Is this behavior intended and should we persist it?

That's why I've used `locator` to disable that flipping for witness function arguments and always match witness agains requirement.

**Comment by aleksgapp** [comment]
> Thanks for working on this! A word of warning -- this is a big change and will require some iteration and a lot more tests. :-) So I hope you don't mind some on-going engagement here before we can merge this in.

Sure, I expected this and challenge accepted :)

> How do we handle ambiguity? With the new approach there might be multiple candidate witnesses for a given requirement. The code in TypeCheckProtocol already has some kind of ranking algorithm but it will need to be generalized.

Yap, the ranking algorithm doesn't work well with new behavior, I'm working on this. 

> With your change, S.req() is now a potential candidate. Are we going to pick it as the witness instead, possibly breaking source compatibility? Or will we pick the extension member which has an exact type?

Despite the fact now we have a choice, still picking the most exact match does make sence, right? As a bonus in this case source compatibility won't be broken.

As for method overrides, I might miss something, but currently it's not possible to make an overriden method with covariant type. If we add this behavior, it will affect only new code, where does source compatibility problem comes from? The new code woudn't be backward compatible, but this is fine, right?

It woud be awesome to make `@objc` protocols and method overrides behavior consistent with protocol conformance, but we can take care of them afterwards.

Thanks for your feedback!

**Comment by xwu** [comment]
> Despite the fact now we have a choice, still picking the most exact match does make sence, right? As a bonus in this case source compatibility won't be broken.

I'd be concerned about that. If the implementation in a conforming type T is enough to satisfy a protocol requirement, then I don't see how one can justify preferring a _default_ implementation. It is inconsistent with the notion of what a default is. Source compatibility can likely be addressed with a clever enough set of warnings.

File: lib/Sema/TypeCheckProtocol.cpp (hunk 5 above)
Comment (slavapestov) [inline lib/Sema/TypeCheckProtocol.cpp]: The second example is correct. Passing a throwing function as a non-throwing function does not work, because the caller does not expect it to throw. If the function does throw (and we admitted this conversion) you would encounter undefined behavior.

Think of `(T) -> U` as a subtype of `(T) throws -> U`, but not the other way around.

**Comment by slavapestov** [comment]
> As for method overrides, I might miss something, but currently it's not possible to make an overriden method with covariant type. 

It is, for a restricted definition of subtyping. Method overrides can make returns more optional (or parameters less optional), they can also make returns into a subclass (or a parameter into a superclass).

However, this case does not work currently:

```
class Base {
  func method() -> SomeProtocol { ... }
}

class Derived : Base {
  override func method() -> SomeConcreteTypeConformingToProtocol { ... }
}
```

If you look at how overrides are matched you'll see it done with custom logic, and not with the constraint solver the way protocol witnesses are. If we get protocol witness matching to support the full range of subtype conversions, we should refactor the code to be usable for method overrides as well.

File: lib/Sema/TypeCheckProtocol.cpp (hunk 6 above)
Comment (slavapestov) [inline lib/Sema/TypeCheckProtocol.cpp]: What was the assert you were getting?

**Comment by martijnwalraven** [comment]
@aleksgapp: Do you think you'll be able to make any progress on this? It would be amazing to get this in in time for Swift 4. I'd have to get familiar with the codebase and brush up on my C++, but happy to help in any way I can!

**Comment by slavapestov** [comment]
I would be wary of getting this into Swift 4 (as we announced the last merge from master into swift-4.0-branch is June 1st) but there is always the next release. Looking forward to more updates here!

**Comment by aleksgapp** [comment]
Sorry, for a long silence. I'm quite busy these days, but hopefully I'll get some free time to work on this.

File: lib/Sema/TypeCheckProtocol.cpp (hunk 6 above)
Comment (aleksgapp) [inline lib/Sema/TypeCheckProtocol.cpp]: `llvm_unreachable("Unhandled transform?")`

**Comment by martijnwalraven** [comment]
@slavapestov: Allowing covariance in protocol conformance would be a big help for our project. Does the approach in this PR still make sense? Or have the relevant parts of the compiler changed in the meantime? 

I don't have much C++ background or knowledge of Swift compiler internals unfortunately, but any pointers on how to get started with contributing to this?

@aleksgapp: Is this still something you're interested in working on?

**Comment by slavapestov** [comment]
@martijnwalraven The approach definitely still makes sense and most of my comments above about implementing this still apply. There are a couple of merge conflicts here but they shouldn't be hard to sort out.

**Comment by martijnwalraven** [comment]
@slavapestov I haven't been able to contribute to this so far unfortunately, but I was wondering if maybe there are efforts underway that you know of to fix this in the Swift 4.1 or Swift 5 timeframe.

**Comment by martijnwalraven** [comment]
@slavapestov I'll likely still won't have time to work on this myself anytime soon, but I was wondering if it is on the core team's roadmap now that ABI stability is done (congratulations on that!). Sorry for not being able to help out here, I know how frustrating it is when users of an open source project keep asking for features without putting in the time themselves :)

File: lib/Sema/TypeCheckProtocol.cpp [hunk 8]
```diff
@@ -1255,9 +1252,16 @@ matchWitness(TypeChecker &tc,
...
+    if (auto func = dyn_cast<FuncDecl>(req)) {
+      auto name = req->getFullName().getBaseName();
```
Comment (shahzadlone) [inline lib/Sema/TypeCheckProtocol.cpp]: should this perhaps be ```const```

**Comment by CodaFi** [comment]
There hasn't been movement here in quite some time.  @aleksgapp please re-open this pull request with a rebased copy of these fixes and tag one of us to review it if you're still interested in pursuing this avenue.
